ouroboros-consensus/src/Ouroboros/Consensus/NodeKernel.hs

```diff
@@ -657,3 +660,19 @@ getMempoolWriter mempool = Inbound.TxSubmissionMempoolWriter
         map (txId . fst) . filter (isMempoolTxAdded . snd) <$>
         addTxs mempool txs
     }
+
+{-------------------------------------------------------------------------------
+  PeerSelection integration
+-------------------------------------------------------------------------------}
+
+-- | Retrieve the peers registered in the current chain/ledger state by
+-- descending stake.
+--
+-- For example, for Shelley, this will return the stake pool relays ordered by
+-- descending stake.
+getPeersFromCurrentLedger ::
+     (IOLike m, LedgerSupportsPeerSelection blk)
+  => NodeKernel m remotePeer localPeer blk
+  -> STM m [(PoolStake, NonEmpty DomainAddress)]
+getPeersFromCurrentLedger kernel =
+    getPeers . ledgerState <$> ChainDB.getCurrentLedger (getChainDB kernel)
```

Comment:
If you care about rollback, the `LedgerState` `k` blocks back should be used instead of the current one. This justifies merging the current and future relays.

Reply:
Take a (`LedgerState blk -> Bool`) predicate. When `False`, return `Nothing`.
In practice the predicate would check whether the ledger state is older or newer than a slot number.